Topic4/src/tests/ShoppingCartTest.java

```diff
@@ -0,0 +1,274 @@
+		Item item4 = new Product("Home Theatre Philips", "HM201", 750.00);
+		Offer offer2 = new Offer("Enjoy your movies", "OF001", 1500.00);
+		offer2.addItem(item3);
+		offer2.addItem(item4);
+		shopcart.addItem(offer2);
+		
+		//offer3 containing two offers
+		Offer offer3 = new Offer("The big offer","OF010",2500.00);
+		offer3.addItem(offer1);
+		offer3.addItem(offer2);
+		shopcart.addItem(offer3);
+
+		List<Item> items = shopcart.getItems();
+		assertEquals(items.size(),3);
+		assertEquals(offer1, items.get(0));
+		assertEquals(offer2, items.get(1));
+		assertEquals(offer3, items.get(2));
+
+		double total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 5150.00);
+	}
+
+	@Test
+	public void addingProductsAndOffers() {
+		ShoppingCart shopcart = new FunctionalShoppingCart();
+
+		// offer1
+		Item item1 = new Product("Smart TV LG 40 inches", "TV100", 700.00);
+		Item item2 = new Product("Home Theatre Sony Muteki 5.1", "HT200", 600.00);
+		Offer offer1 = new Offer("Cinema in your living", "OF001", 1150.00);
+		offer1.addItem(item1);
+		offer1.addItem(item2);
+		shopcart.addItem(offer1);
+		assertEquals(shopcart.getItems().size(), 1);
+		List<Item> items = shopcart.getItems();
+		assertEquals(offer1, items.get(0));
+
+		// product1
+		Item item3 = new Product("Table TV O", "TT001", 100.00);
+		shopcart.addItem(item3);
+		assertEquals(shopcart.getItems().size(), 2);
+
+		items = shopcart.getItems();
+		assertEquals(offer1, items.get(0));
+		assertEquals(item3, items.get(1));
+		double total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 1250.00);
+	}
+
+	@Test
+	public void removingItems() {
+
+		ShoppingCart shopcart = new FunctionalShoppingCart();
+
+		// offer1
+		Item item1 = new Product("Smart TV LG 40 inches", "TV100", 700.00);
+		Item item2 = new Product("Home Theatre Sony Muteki 5.1", "HT200", 600.00);
+		Offer offer1 = new Offer("Cinema in your living", "OF001", 1150.00);
+		offer1.addItem(item1);
+		offer1.addItem(item2);
+		shopcart.addItem(offer1);
+
+		// product1
+		Item item3 = new Product("Table TV O", "TT001", 100.00);
+		shopcart.addItem(item3);
+
+		assertEquals(shopcart.getItems().size(), 2);
+		List<Item> items = shopcart.getItems();
+		assertEquals(offer1, items.get(0));
+		assertEquals(item3, items.get(1));
+
+		double total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 1250.00);
+
+		shopcart.removeItem(offer1);
+		assertEquals(shopcart.getItems().size(), 1);
+		items = shopcart.getItems();
+		assertEquals(item3, items.get(0));
+
+		total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 100.00);
+
+		shopcart.removeItem(item3);
+		assertEquals(shopcart.getItems().size(), 0);
+
+		total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 0.00);
+	}
+
+	@Test
+	public void addingRemovingPayment() {

```

Comment:
By TDD definition it should be splitted in two test, one to test Adding and another to test Removing